### ISSUE TITLE
Consistent extend-package command

### ISSUE THREAD

**Issue by tabacitu** [Laravel-Backpack/Base#131]
I think the next big improvement in the Backpack packages will be to make it easier to extend them, and let the developer modify all classes to his liking. Right now that's a little difficult, and we don't have documentation for each package on how to do that. This proposal aims to bring a consistent way to extend all Backpack packages.

Packages involved (and what will be extended):
- Base (DashboardController, Auth controllers, admin middleware, ResetPasswordNotification)
- CRUD (CrudController, CrudRequest, CrudPanel)
- LogManager (LogController)
- PageManager (PageCrudController, PageRequest, Models\Page)
- PermissionManager (controllers, requests, models)
- NewsCrud (controllers, requests, models)
- MenuCrud (controller, request, model)

Commands:
- php artisan backpack:extend-base
- php artisan backpack:extend-crud
- php artisan backpack:extend-logmanager
- php artisan backpack:extend-pagemanager
- php artisan backpack:extend-permissionmanager
- php artisan backpack:extend-newscrud
- php artisan backpack:extend-menucrud

Each command would:
- publish blank controllers in ```app/Http/Controllers/Admin/``` that extend the ones in the package;
- publish blank models in ```app/Models/``` that extend the ones in the package;
- publish blank requests in ```app/Requests/``` that extend the ones in the package;
- publish blank middleware in ```app/Middleware/``` that extend the ones in the package;
- publish blank php classes in ```app/Library/``` that extend the other classes in the packages (CrudPanel for example);
- create a routes file inside ```routes/backpack``` with the routes pointing to the new controllers;
- echo in the command line any more steps that the developer needs to do to finish extending the package (no need to look over the documentation to find out);

File naming:
- the extended files will have the exact same name as the ones they extend; the only difference will be the namespace (one is inside ```App/Http```, one is inside ```Backpack/PageManager/app/Http/etc```).

Result:
Basically running an extend command will publish all the blank classes you need to extend a package, and the routes file to make the application use those new files. For most packages, this single command is all you need to do.

Possible problems:
- pagemanager has its classes defined in the config file; in order to make extending easier, those config options will go away;

Did I miss anything? Do you guys see any problems I haven't in implementing this? Thanks, cheers!

**Comment by lloy0076** [comment]
So long as the publish honours stuff in the configuration such as the 'base' (usually /admin but it is configurable).

This functionality would be good to have as I've extended `CrudController` and `CrudPanel`; the first to handle things that `$request->expectsJson()` returns `true` to and the latter because I want to intercept some of the panel's functionality.

**Comment by eduardoarandah** [comment]
Yes! 

However, I feel this goes from easy to super complex very fast. 

I feel that custom routing solves most common issues

- Check permissions to manage users? add middleware('can:manage-users')
- Want to log some things? create middleware and add it to routes
- Want to add middleware? add it to routes
- Want to remove functionality? remove it in routes

For the intermediate:

- Want to change package controller? manually create extended controller and point to it in routes file
- Want to add more validations to every request? in extended controller

Advanced:

- Extending CrudPanel
- Extending CrudController
- Extending Library

Regarding Package Middleware, I feel it doesn't even have to exist. Laravel has it's own 'auth' middleware with same functionality than 'admin' middleware. 
For new functionality, it's easier to just create your own, as usual

**Comment by tabacitu** [comment]
Hmm... you may be onto something very very nice, @eduardoarandah . I love it when one solution fixes multiple problems. Like money [I heard :-)) ]

Extending routes is definitely a must. So let's do _that_ first (noticed you already submitted PRs) and see if it's actually NEEDED to build upon it, to offer these extend commands for every little thing. Baby steps :-)

Let's leave this open until then, it'll also provide a place for others to give feedback.

Cheers!

**Comment by eduardoarandah** [comment]
@tabacitu I'm on a project right now and the backpack routes file felt completely natural. 

Most of the time I'm CTRL+P in sublime
jumping from routes->controllers->views

For simple entities, all it takes is 
- adding a couple of methods to backpack controller
- add the view
- validation
- permission via middleware 
- ready!

**Comment by tabacitu** [comment]
![giphy](https://user-images.githubusercontent.com/1032474/28320983-ad9a0cea-6bda-11e7-9025-4f5ced56f7ba.gif)

**Comment by tabacitu** [comment]
Guys, I'm very happy with how extensible Backpack is right now, and it looks like the direction we're headed is decentralized packages, rather than centralized (thanks in part to people like @eduardoarandah who create compatible packages or alternative solutions ;-) ).

In the universe where each package is autonomous, but all depend on Backpack\Base and maybe Backpack\CRUD, I don't see this extend command as "a feature", but rather as a "guideline" that will help people who build packages, to make it easier for users to extend them. 

That being said, something I haven't thought of before is that it's a fine line... If we make it too easy, people might extend them even when they don't _need to_, and by doing that prevent themselves from benefiting from updates... Food for thought... 

Since this has been greatly improved by the "custom routes" feature in all packages (thanks again @eduardoarandah ), I say we table the issue for now, and come back to it in case we find it too uncomfortable to extend in the future.

Cheers!

**Comment by eduardoarandah** [comment]
(Sorry if this comment reopens the issue) 

My thoughts: 

I've made a LOT of wordpress sites before. 
Dealing with users with zero experience I found some things:

Wordpress base functionality covers 90% of all requirements for a user interface in the backend. 

Comparing it with backpack, there's only a couple of things missing: 
A **super-easy** media manager capable of "attaching" its images to other entities, with auto-resize-crop

![clipboard01](https://user-images.githubusercontent.com/4065733/30282690-a3c4bcb6-96db-11e7-9e25-cd6f0dbae9ba.png)

This could be covered with this package, haven't tried it but seems good

https://murze.be/2017/08/getting-started-laravel-medialibrary-video-tutorial/

A dashboard with draggable items.

![fff](https://user-images.githubusercontent.com/4065733/30282861-1d5a25ca-96dc-11e7-9676-6dc623140781.png)

And finally, a "centralized" way of updating settings for the plugins installed. 

In this example, the "Akismet" plugin automatically created a settings->akismet screen. 

![rrr](https://user-images.githubusercontent.com/4065733/30283080-b9c408e0-96dc-11e7-8700-ca953c7ec0a0.png)

The settings screen is nothing more than buttons and fields
![sett](https://user-images.githubusercontent.com/4065733/30283666-494943ee-96de-11e7-8cb0-88cc1cf506d6.png)


This is crucial because: 
1) You can't ask the user to edit .env files or config/myplugin.php files 
2) You can't ask the user to pay a programmer every time they want to change some setting
3) You can't put eveything in Backpack/Settings, it gets confusing
4) You want end users being able to change its google analytics code, but deny other settings

There must be a plug-and-play way of letting the developer "register" its settings in 
backpack-settings (to create a settings page)
backpack-base (to create a menu item to access that settings page)

Maybe namespacing settings? 
config('settings.myplugin.analytics-code')

Anyway, this is just my thoughts coming from a lot of years doing wordpress 

Another random thought: 

I once tried OctoberCMS but didn't like it because it adds complexity instead of removing it, the worst of all sins for the [KISS god](https://en.wikipedia.org/wiki/KISS_principle)

Backpack removes complexity when it's not needed.
Saves you from creating a bunch of routes, views and requests. You only code them when needed.

have a good day!

**Comment by tabacitu** [comment]
@eduardoarandah totally agree with the extra packages/features needed for Backpack to compete with WordPress (at least for simple blogs at first):
- installation wizard
- media library (you don't think elFinder is enough for that?)
- dashboard with an API/settings for packages to make use of it
- settings package with tabs and an API (I agree namespacing should solve this)
- sidebar/top menu with an API

I'd also add:
- package that allows you to install&update other packages in production (as in, run ```composer install``` and ```composer update``` in production);

Also agree on OctoberCMS. Tried it before starting Backpack, it's sad to see how difficult it makes your life, they have a lot of good stuff in there.
